Tutorial: How to set up Next.js error monitoring (#10958)
PostHog/posthog.com: contents/tutorials/nextjs-error-monitoring.md

To help you set this up, this tutorial details how to create a basic Next.js app, set up PostHog on both the front and backend, and then automatically capture errors that happen in both locations.  

## 1. Creating a Next.js app

Start by ensuring [Node.js is installed](https://nodejs.dev/en/learn/how-to-install-nodejs/) (version 18.0 or newer) then run the following command. Say **no** to TypeScript, **yes** to app router, and the defaults for other options.

```bash
npx create-next-app@latest next-errors
```

Next, we can create our frontend which will have two parts:

1. A button that throws an error
2. A button that makes a request to our backend API

We can modify `app/page.js` to do this:

```js
'use client'

import styles from "./page.module.css";

export default function Home() {

  const handleErrorButtonClick = () => {
    throw new Error("Frontend error");
  }

  const handleAPIButtonClick = async () => {
    const response = await fetch("/api/test-error");
    const data = await response.json();
    console.log("data", data);
  }

  return (
    <div className={styles.page}>

>       <h1>Welcome to our broken app</h1>

Reviewer: lol

/>

### Backend setup

For the backend, we can create a `posthog-server.js` file in the `app` directory. In it, initialize PostHog from `posthog-node` as a singleton with your project API key and host from [your project settings](https://us.posthog.com/settings/project). This looks like this:

```js
// app/posthog-server.js
import { PostHog } from 'posthog-node'

let posthogInstance = null

export function getPostHogServer() {
  if (!posthogInstance) {
    posthogInstance = new PostHog(
      '<ph_project_api_key>', 
      {
        host: '<ph_client_api_host>',
        flushAt: 1,
        flushInterval: 0
      }
    )
  }
  return posthogInstance
}
```

We can then import this singleton wherever we need it in the backend. Unfortunately, this doesn’t autocapture errors by default, so we have some more work to do.

## 3. Capturing errors

With both front and backend initializations set up, capturing errors with PostHog is as simple as calling `captureException` or capturing an `$exception` event. 

<MultiLanguage>

```js
posthog.captureException(error, additionalProperties)
```

```node
posthog.captureException(e, 'user_distinct_id', additionalProperties)
```

</MultiLanguage>

> Doing this for every possible error is a hassle though and you’ll inevitably miss errors we’re not expecting. Our frontend implementation automatically captures errors thrown and caught by `onError` and `onUnhandledRejection` listeners, but this doesn’t cover everything.

Reviewer: ```suggestion
Doing this for every possible error is a hassle though and we'll inevitably miss errors we’re not expecting. Our frontend implementation automatically captures errors thrown and caught by `onError` and `onUnhandledRejection` listeners, but this doesn’t cover everything.
```

This looks like this:

```js
// instrumentation.js
export function register() {
  // No-op for initialization
}

export const onRequestError = async (err, request, context) => {
  if (process.env.NEXT_RUNTIME === 'nodejs') {
    const { getPostHogServer } = require('./app/posthog-server')
    const posthog = await getPostHogServer()

    let distinctId = null
    if (request.headers.cookie) {
      const cookieString = request.headers.cookie
      const postHogCookieMatch = cookieString.match(/ph_phc_.*?_posthog=([^;]+)/)
      
      if (postHogCookieMatch && postHogCookieMatch[1]) {
        try {
          const decodedCookie = decodeURIComponent(postHogCookieMatch[1])
          const postHogData = JSON.parse(decodedCookie)
          distinctId = postHogData.distinct_id
        } catch (e) {
          console.error('Error parsing PostHog cookie:', e)
        }
      }
    }

    await posthog.captureException(err, distinctId || undefined)
  }
}
```

> Now, when you click the **Click me for a backend API error** button, it will trigger and error, and that error will be automatically captured by PostHog. 

Reviewer: ```suggestion
Now, when you click the **Click me for a backend API error** button, it will trigger an error which will be automatically captured by PostHog. 
```

<ProductScreenshot
  imageLight="https://res.cloudinary.com/dmukukwp6/image/upload/Clean_Shot_2025_03_14_at_15_30_05_2x_e99f29a546.png"
  imageDark="https://res.cloudinary.com/dmukukwp6/image/upload/Clean_Shot_2025_03_14_at_15_30_23_2x_c4f513e4a7.png"
  alt="PostHog"
  classes="rounded"
/>

## 4. Monitoring errors in PostHog

> Once you’ve set up error capture in your app, you can head to the [error tracking tab](https://us.posthog.com/error_tracking) in PostHog to few the issues popping up along with their frequency.

Reviewer: ```suggestion
Once you’ve set up error capture in your app, you can head to the [error tracking tab](https://us.posthog.com/error_tracking) in PostHog to review the issues popping up along with their frequency.
```
